src/software/ai/navigator/obstacle/obstacle_factory.cpp

```diff
@@ -1,101 +1,193 @@
+    {
+        // use hexagonal approximation for velocity obstacle
+        Vector velocity_direction_norm_radius =
+            velocity_cushion_vector.normalize(radius_cushion);
+        return createObstacle(Polygon(
+            {// left side of robot
+             robot.position() + velocity_direction_norm_radius.rotate(Angle::quarter()),
+             // back left of robot
+             robot.position() +
+                 velocity_direction_norm_radius.rotate(Angle::fromDegrees(150)),
+             // back right of robot
+             robot.position() +
+                 velocity_direction_norm_radius.rotate(Angle::fromDegrees(210)),
+             // right side of robot
+             robot.position() +
+                 velocity_direction_norm_radius.rotate(Angle::threeQuarter()),
+             // right side velocity cushions
+             robot.position() +
+                 velocity_direction_norm_radius.rotate(Angle::threeQuarter()) +
+                 velocity_cushion_vector,
+             // left side velocity cushions
+             robot.position() + velocity_direction_norm_radius.rotate(Angle::quarter()) +
+                 velocity_cushion_vector}));
+    }
+    else
+    {
+        return createObstacle(Circle(robot.position(), radius_cushion));
+    }
 }
 
-std::vector<Obstacle> ObstacleFactory::getVelocityObstaclesFromTeam(const Team &team)
+std::vector<Obstacle> ObstacleFactory::createVelocityObstaclesFromTeam(const Team &team)
 {
     std::vector<Obstacle> obstacles;
     for (const auto &robot : team.getAllRobots())
     {
-        obstacles.push_back(getVelocityObstacleFromRobot(robot));
+        obstacles.push_back(createVelocityObstacleFromRobot(robot));
     }
     return obstacles;
 }
+
+Obstacle ObstacleFactory::createBallObstacle(const Point &ball_position)
+{
+    return createObstacle(
+        Circle(ball_position, BALL_MAX_RADIUS_METERS + 0.06 + shape_expansion_amount));
+}
+
+Obstacle ObstacleFactory::createRobotObstacle(const Point &robot_position,
+                                              const double radius_scaling)
+{
+    return createObstacle(
+        Circle(robot_position,
+               radius_scaling * (ROBOT_MAX_RADIUS_METERS + shape_expansion_amount)));
+}
+
+Obstacle ObstacleFactory::createObstacleFromRectangle(const Rectangle &rectangle)
+{
+    Rectangle rectangle_exp(rectangle);
+    rectangle_exp.expand(shape_expansion_amount);
+    return createObstacle(rectangle_exp);
+}
+
+Obstacle ObstacleFactory::createObstacle(const Circle &circle)
+{
+    return Obstacle(std::make_shared<CircleObstacle>(CircleObstacle(circle)));
+}
+
+Obstacle ObstacleFactory::createObstacle(const Polygon &polygon)
+{
+    return Obstacle(std::make_shared<PolygonObstacle>(PolygonObstacle(polygon)));
+}
+
+std::optional<Rectangle> ObstacleFactory::getRectangleFromRectangularMotionConstraint(
+    const MotionConstraint &motion_constraint, const World &world)
+{
+    Rectangle rectangle({0, 0}, {1, 1});
```

Comment:
This still feels like a weird way of declaring a default/fallback value. Since there's an early return in the `default` case, this default value is never used unless we totally forget to override it when handling a case (in which case there's no warning and we just use this random value).

My suggestion would be to declare this variable as an optional rectangle. Then the switch statement can stay basically the same, but the default case could just explicitly supply a nullopt. Then you can check the optional before doing the final expand, then return whatever the value is. I think avoiding mixing early returns with a single return at the end would help make things a bit more clear.

Reply:
I think we should remove the `default` case entirely. We really should be explicitly handling each case.

Reply:
I think a default case is fine to warn if we aren't handling something, I just think what we're doing currently is weird. If for whatever reason the compiler stopped warning about unhandled enums in the case statement it's good to have something there.

Reply:
hm I struggled a bit with this but I agree that it should all be in the switch statement in the  createObstacleFromMotionConstraint function